src/avram/model.cr

```diff
@@ -184,6 +184,79 @@ class Avram::Model
     {% COLUMNS << {name: type_declaration.var, type: data_type, nilable: nilable.id, autogenerated: autogenerated} %}
   end
 
+  macro polymorphic_belongs_to(type_declaration, autogenerated = false)
+
+    {% nilable = false %}
+    {% for type in type_declaration.type.types %}
+      {% if type.class_name == "Path" %}
+        {% nilable = true if type.names[0] == "Nil" %}
+      {% elsif type.class_name == "Union" %}
+        {% nilable = true if type.types.any? { |item| item.names[0] == "Nil" } %}
+      {% end %}
+    {% end %}
+
+    {% COLUMNS << {name: type_declaration.var + "_id", type: Int64, nilable: nilable.id, autogenerated: autogenerated} %}
+    {% COLUMNS << {name: type_declaration.var + "_type", type: String, nilable: nilable.id, autogenerated: autogenerated} %}
+
+    macro finished
+      class SaveOperation
+        before_save :validates_polymorphic_association
+
+        private def validates_polymorphic_association
+          previous_def
+
+          {% if nilable %}
+          validate_inclusion_of {{ type_declaration.var }}_type, in: {{type_declaration.type.types.map(&.stringify).push("").push(nil)}}
+          {% else %}
+          validate_required {{ type_declaration.var }}_type
+          validate_required {{ type_declaration.var }}_id
+          validate_inclusion_of {{ type_declaration.var }}_type, in: {{type_declaration.type.types.map(&.stringify)}}
+          {% end %}
+        end
+      end
+    end
+
+    def {{type_declaration.var}} : {{type_declaration.type}}{% if nilable %}?{% end %}
+    {% if nilable %}
+      return nil if {{type_declaration.var}}_id.nil?
+    {% end %}
+
+    {% if type_declaration.type.is_a?(Union) %}
+      {% first = true %}
```

Comment:
Can you explain what this is for? And/or maybe add some comments on how it is used?

Reply:
Yeah, this is because the macro generates this code:

```
  27.     def commentable : Employee | Company
  28.     
  29. 
  30.     
  31.       
  32.       
  33.         
  34.       if commentable_type == "Employee"
  35.         commentable_id.try do |lookup_id|
  36.           Employee::BaseQuery.find(lookup_id)
  37.         end
  38.         
  39.       
  40.       
  41.         
  42.       elsif commentable_type == "Company"
  43.         commentable_id.try do |lookup_id|
  44.           Company::BaseQuery.find(lookup_id)
  45.         end
  46.         
  47.       
  48.       else
  49.         raise "Unknown commentable: #{commentable_type}"
  50.       end
  51.       
  52.       
  53.     
  54.     end
```

So for the first element, it's an `if`, for all subsequent, it is an `elseif` that why I use this `first` variable, any simpler way of doing this? should I add a comment?

Reply:
I believe you can use `case` for this to simplify this so it generates something like:

```crystal
case commentbale_type
when "Company"
  # do stuff
when "Employee"
  # do stuff
else
  # raise
end
```

That way it works for any number of items

Reply:
you're totally right, way better heheh